components/proxy/src/main/java/com/hotels/styx/proxy/backends/file/FileChangeMonitor.java

```diff
@@ -0,0 +1,155 @@
+import com.google.common.hash.HashCode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hotels.styx.common.Files.fileContentMd5;
+import static java.lang.String.format;
+import static java.nio.file.Files.exists;
+import static java.nio.file.Files.getLastModifiedTime;
+import static java.nio.file.Files.isReadable;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Monitors a file system object and notifies the consumer of any changes.
+ */
+public class FileChangeMonitor implements FileMonitor {
+    private final Path monitoredFile;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long pollPeriod;
+    private final TimeUnit timeUnit;
+    private ScheduledFuture<?> monitoredTask;
+
+    private AtomicReference<FileTime> lastChangedTime = new AtomicReference<>(FileTime.fromMillis(0));
+    private AtomicReference<HashCode> hashCode = new AtomicReference<>();
+    private volatile boolean performHashCheck;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileChangeMonitor.class);
+
+    @VisibleForTesting
+    FileChangeMonitor(String monitoredFile, long pollPeriod, TimeUnit timeUnit) {
+        requireExists(requireNonNull(monitoredFile));
+        this.monitoredFile = Paths.get(monitoredFile);
+        this.pollPeriod = pollPeriod;
+        this.timeUnit = requireNonNull(timeUnit);
+        this.hashCode.set(HashCode.fromLong(0));
+    }
+
+    public FileChangeMonitor(String monitoredFile) {
+        this(monitoredFile, 1, SECONDS);
+    }
+
+    @Override
+    public void start(Listener listener) {
+        synchronized (this) {
+            if (monitoredTask != null) {
+                throw new IllegalStateException(format("File monitor for '{}' is already started", monitoredFile));
+            }
+
+            monitoredTask = executor.scheduleAtFixedRate(newPollerTask(listener), pollPeriod, pollPeriod, timeUnit);
+        }
+    }
+
+    private Runnable newPollerTask(Listener listener) {
+        return () -> {
+            if (!exists(monitoredFile)) {
+                LOGGER.debug("Monitored file does not exist. Path={}", monitoredFile);
+                return;
+            }
+
+            if (!isReadable(monitoredFile)) {
+                LOGGER.debug("Monitored file is no longer readable. Path={}", monitoredFile);
+                return;
+            }
+
+            if (modificationTimeChanged(monitoredFile)) {
```

Comment:
We might want to refactor this into a new method " fileHasChanged" . 

Potentially, we might want to include a "file size" check just after the timestamp; that way, we could avoid having to perform the md5sum over the whole content when the file size has changed.

Reply:
In principle I think it would be a good idea. However in this instance I would caution against it. For the following reasons:

1) The hash code is computed only once on the next poll cycle, regardless of the file size change. This is to ensure we always pick up the latest changes if the file changes incrementally.

2) It adds little bit more code which is really quite tricky to test.

For these reasons I would prefer for the simpler option and to keep the length it out.